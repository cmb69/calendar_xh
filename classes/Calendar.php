<?php

/**
 * Copyright 2021 Christoph M. Becker
 *
 * This file is part of Calendar_XH.
 *
 * Calendar_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Calendar_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Calendar_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Calendar;

class Calendar
{
    /** @var bool */
    private $weekStartsMonday;

    /**
     * @param bool $weekStartsMonday
     */
    public function __construct($weekStartsMonday)
    {
        $this->weekStartsMonday = $weekStartsMonday;
    }

    /**
     * @param int $year
     * @param int $month
     * @return array<int,array<int,(int|null)>>
     */
    public function getMonthMatrix($year, $month)
    {
        $result = [];
        $row = [];
        $days = (int) date('t', mktime(1, 1, 1, $month, 1, $year));
        for ($day = 1; $day <= $days; $day++) {
            $dayofweek = $this->getDayOfWeek($year, $month, $day);
            if ($day === 1) {
                for ($i = 0; $i < $dayofweek; $i++) {
                    $row[] = null;
                }
            }
            $row[] = $day;
            if ($day === $days) {
                for ($i = $dayofweek + 1; $i < 7; $i++) {
                    $row[] = null;
                }
            }
            if (count($row) === 7 || $day === $days) {
                $result[] = $row;
                $row = [];
            }
        }
        return $result;
    }

    /**
     * @param int $year
     * @param int $month
     * @param int $day
     * @return int
     */
    private function getDayOfWeek($year, $month, $day)
    {
        $dayofweek = (int) date('w', mktime(1, 1, 1, $month, $day, $year));
        if ($this->weekStartsMonday) {
            $dayofweek = ($dayofweek + 6) % 7;
        }
        return $dayofweek;
    }
}
