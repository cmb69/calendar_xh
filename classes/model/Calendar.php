<?php

/**
 * Copyright (c) Christoph M. Becker
 *
 * This file is part of Calendar_XH.
 *
 * Calendar_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Calendar_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Calendar_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Calendar\Model;

use Calendar\Html2Text;

class Calendar
{
    /** @var array<string,Event> */
    private $events;

    /** @param array<string,Event> $events */
    public function __construct(array $events)
    {
        $this->events = $events;
    }

    /** @return array<string,Event> */
    public function events(): array
    {
        return $this->events;
    }

    public function event(string $id): ?Event
    {
        return $this->events[$id] ?? null;
    }

    public function nextEvent(LocalDateTime $now): ?Event
    {
        $nextevent = null;
        $nextldt = null;
        foreach ($this->events as $event) {
            $ldt = $event->after($now);
            if ($ldt === null) {
                continue;
            }
            if ($nextldt === null || $ldt->compare($nextldt) < 0) {
                $nextevent = $event;
                $nextldt = $ldt;
            }
        }
        return $nextevent;
    }

    /** @return list<Event> */
    public function eventsDuring(int $year, int $month): array
    {
        $result = [];
        foreach ($this->events as $event) {
            if ($event->occursDuring($year, $month)) {
                $result[] = $event;
            }
        }
        uasort($result, function (Event $a, Event $b) use ($year): int {
            $dt1 = $a->isBirthday() ? $a->start()->withYear($year) : $a->start();
            $dt2 = $b->isBirthday() ? $b->start()->withYear($year) : $b->start();
            return $dt1->compare($dt2);
        });
        return $result;
    }

    /** @return list<Event> */
    public function eventsOn(LocalDateTime $day, bool $daysBetween): array
    {
        assert($day->hour() === 0 && $day->minute() === 0);
        $result = [];
        foreach ($this->events as $event) {
            if ($event->occursOn($day, $daysBetween)) {
                $result[] = $event;
            }
        }
        return $result;
    }

    public function delete(string $id): void
    {
        assert(array_key_exists($id, $this->events));
        unset($this->events[$id]);
    }

    public function toICalendarString(Html2Text $converter, string $host): string
    {
        $res = "BEGIN:VCALENDAR\r\n"
            . "PRODID:-//3-magi.net//Calendar_XH//EN\r\n"
            . "VERSION:2.0\r\n";
        foreach ($this->events() as $id => $event) {
            $res .= $event->toICalendarString($id, $converter, $host);
        }
        $res .= "END:VCALENDAR\r\n";
        return $res;
    }
}
