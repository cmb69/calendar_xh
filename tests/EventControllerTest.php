<?php

namespace Calendar;

use ApprovalTests\Approvals;
use Calendar\Infra\DateTimeFormatter;
use Calendar\Model\Calendar;
use Calendar\Model\Event;
use Calendar\Model\LocalDateTime;
use Calendar\Model\NoRecurrence;
use Calendar\Model\YearlyRecurrence;
use org\bovigo\vfs\vfsStream;
use PHPUnit\Framework\TestCase;
use Plib\DocumentStore;
use Plib\FakeRequest;
use Plib\View;

class EventControllerTest extends TestCase
{
    /** @var array<string,string> */
    private $config;

    /** @var DocumentStore */
    private $store;

    /** @var DateTimeFormatter */
    private $dateTimeFormatter;

    /** @var View */
    private $view;

    public function setUp(): void
    {
        vfsStream::setup("root");
        $this->config = XH_includeVar("./config/config.php", "plugin_cf")["calendar"];
        $this->store = new DocumentStore(vfsStream::url("root/"));
        $calendar = Calendar::updateIn($this->store);
        $calendar->addEvent("111", $this->lunchBreak()->toDto());
        $calendar->addEvent("222", $this->christmas()->toDto());
        $this->store->commit();
        $lang = XH_includeVar("./languages/en.php", "plugin_tx")["calendar"];
        $this->dateTimeFormatter = new DateTimeFormatter($lang);
        $this->view = new View("./views/", $lang);
    }

    private function sut(): EventController
    {
        return new EventController(
            $this->config,
            $this->store,
            $this->dateTimeFormatter,
            $this->view,
        );
    }

    public function testDoesNothingWhenNotEnabled(): void
    {
        $request = new FakeRequest(["url" => "http://example.com/?&function=calendar_event&event_id=111"]);
        $response = $this->sut()($request);
        $this->assertSame("", $response->output());
    }

    public function testReportsNotFoundForUnknownEvent(): void
    {
        $this->config["event_allow_single"] = "true";
        $request = new FakeRequest(["url" => "http://example.com/?&function=calendar_event&event_id=unknown"]);
        $response = $this->sut()($request);
        $this->assertSame(404, $response->status());
    }

    public function testRendersKnownEvent(): void
    {
        $this->config["event_allow_single"] = "true";
        $request = new FakeRequest(["url" => "http://example.com/?&function=calendar_event&event_id=111"]);
        $response = $this->sut()($request);
        Approvals::verifyHtml($response->output());
    }

    public function testTruncatesLongDescription(): void
    {
        $this->config["event_allow_single"] = "true";
        $request = new FakeRequest(["url" => "http://example.com/?&function=calendar_event&event_id=222"]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("The kitchen smells from â€¦", $response->output());
    }

    private function lunchBreak(): Event
    {
        $start = new LocalDateTime(2023, 1, 4, 12, 0);
        $end = new LocalDateTime(2023, 1, 4, 13, 0);
        $recurrence = new NoRecurrence($start, $end);
        return new Event("", $start, $end, "Lunch break", "", "Tips for lunch breaks", "whereever I am", $recurrence);
    }

    private function christmas(): Event
    {
        $start = new LocalDateTime(2020, 12, 25, 0, 0);
        $end = new LocalDateTime(2020, 12, 26, 23, 59);
        $recurrence = new YearlyRecurrence($start, $end, new LocalDateTime(2030, 12, 25, 0, 0));
        $description = "Christmas is a wonderful time of the year."
            . " Everything is silent and white. Everybody is peaceful."
            . " The snow falls silently on the lake."
            . " The kitchen smells from fresh baked cake.";
        return new Event("222", $start, $end, "Christmas", "", $description, "Tree", $recurrence);
    }
}
