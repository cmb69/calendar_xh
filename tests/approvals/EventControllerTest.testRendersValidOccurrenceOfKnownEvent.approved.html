
<article class="calendar_event">
  <h1>Christmas</h1>
  <p class="event_date_time"><span class="event_date">12/25/2025</span> till <span class="event_date">12/26/2025</span></p>
  <p class="event_location">Tree</p>
  <div class="event_description">Christmas is a wonderful time of the year. Everything is silent and white. Everybody is peaceful. The snow falls silently on the lake. The kitchen smells from fresh baked cake.</div>
  <script type="application/ld+json">{"@context":"https://schema.org","@type":"Event","name":"Christmas","startDate":"2025-12-25","duration":"P1DT23H59M","url":"http://example.com/?&function=calendar_event&event_id=222&calendar_occurrence=2025-12-25","location":"Tree","description":"Christmas is a wonderful time of the year. Everything is silent and white. Everybody is peaceful. The snow falls silently on the lake. The kitchen smells from â€¦"}</script>
</article>
